File: experiments-local/experiment-mu-sonnet-claudecode/runs/agent=claude-code_language=typescript_model=sonnet-4.6_tooling=none/rep3/src/tools/team-tools.ts
import { getDataStore, teamMatches } from "../data-loader.js";
import type { Match, TeamStats } from "../types.js";

function computeStats(matches: Match[], teamName: string): TeamStats {
  let wins = 0, draws = 0, losses = 0, goalsFor = 0, goalsAgainst = 0;

  for (const m of matches) {
    const isHome = teamMatches(m.homeTeam, teamName);
    const gf = isHome ? m.homeGoals : m.awayGoals;
    const ga = isHome ? m.awayGoals : m.homeGoals;
    goalsFor += gf;
    goalsAgainst += ga;
    if (gf > ga) wins++;
    else if (gf === ga) draws++;
    else losses++;
  }

  return {
    team: teamName,
    matches: matches.length,
    wins,
    draws,
    losses,
    goalsFor,
    goalsAgainst,
    points: wins * 3 + draws,
  };
}

export function getTeamStats(params: {
  team: string;
  competition?: string;
  season?: number;
  homeOnly?: boolean;
  awayOnly?: boolean;
}): string {
  const { matches } = getDataStore();

  let filtered = matches.filter((m) => {
    const isHome = teamMatches(m.homeTeam, params.team);
    const isAway = teamMatches(m.awayTeam, params.team);
    if (!isHome && !isAway) return false;
    if (params.homeOnly && !isHome) return false;
    if (params.awayOnly && !isAway) return false;
    if (params.competition) {
      const comp = params.competition.toLowerCase();
      if (comp.includes("brasileirao") && m.competition !== "brasileirao" && m.competition !== "historical") return false;
      if (comp.includes("copa do brasil") && m.competition !== "copa_do_brasil") return false;
      if (comp.includes("libertadores") && m.competition !== "libertadores") return false;
    }
    if (params.season && m.season !== params.season) return false;
    return true;
  });

  if (filtered.length === 0) {
    return `No matches found for ${params.team} with the given filters.`;
  }

  const stats = computeStats(filtered, params.team);
  const winRate = stats.matches > 0 ? ((stats.wins / stats.matches) * 100).toFixed(1) : "0.0";
  const avgGoalsFor = stats.matches > 0 ? (stats.goalsFor / stats.matches).toFixed(2) : "0.00";
  const avgGoalsAgainst = stats.matches > 0 ? (stats.goalsAgainst / stats.matches).toFixed(2) : "0.00";
  const goalDiff = stats.goalsFor - stats.goalsAgainst;

  const context = [
    params.competition ?? "All competitions",
    params.season ? `Season ${params.season}` : null,
    params.homeOnly ? "Home only" : params.awayOnly ? "Away only" : null,
  ]
    .filter(Boolean)
    .join(", ");

  return [
    `${stats.team} Statistics (${context})`,
    `Matches: ${stats.matches}`,
    `Record: ${stats.wins}W ${stats.draws}D ${stats.losses}L`,
    `Points: ${stats.points}`,
    `Goals For: ${stats.goalsFor} (avg ${avgGoalsFor}/match)`,
    `Goals Against: ${stats.goalsAgainst} (avg ${avgGoalsAgainst}/match)`,
    `Goal Difference: ${goalDiff >= 0 ? "+" : ""}${goalDiff}`,
    `Win Rate: ${winRate}%`,
  ].join("\n");
}

export function getStandings(params: { season: number; competition?: string }): string {
  const { matches } = getDataStore();

  let filtered = matches.filter((m) => {
    if (m.season !== params.season) return false;
    if (params.competition) {
      const comp = params.competition.toLowerCase();
      if (comp.includes("brasileirao") || comp.includes("serie a")) {
        if (m.competition !== "brasileirao" && m.competition !== "historical") return false;
      } else if (comp.includes("copa do brasil")) {
        if (m.competition !== "copa_do_brasil") return false;
      } else if (comp.includes("libertadores")) {
        if (m.competition !== "libertadores") return false;
      }
    } else {
      // Default to league matches
      if (m.competition !== "brasileirao" && m.competition !== "historical") return false;
    }
    return true;
  });

  if (filtered.length === 0) {
    return `No matches found for season ${params.season}.`;
  }

  const teamMap = new Map<string, { matches: Match[]; name: string }>();

  for (const m of filtered) {
    for (const team of [m.homeTeam, m.awayTeam]) {
      const key = team.toLowerCase();
      if (!teamMap.has(key)) teamMap.set(key, { matches: [], name: team });
      teamMap.get(key)!.matches.push(m);
    }
  }

  const table: TeamStats[] = [];
  for (const [, entry] of teamMap) {
    table.push(computeStats(entry.matches, entry.name));
  }

  table.sort((a, b) => {
    if (b.points !== a.points) return b.points - a.points;
    const gdB = b.goalsFor - b.goalsAgainst;
    const gdA = a.goalsFor - a.goalsAgainst;
    if (gdB !== gdA) return gdB - gdA;
    return b.goalsFor - a.goalsFor;
  });

  const compLabel = params.competition ?? "Brasileirão";
  const lines = table.slice(0, 20).map((t, i) => {
    const gd = t.goalsFor - t.goalsAgainst;
    const gdStr = (gd >= 0 ? "+" : "") + gd;
    return `${String(i + 1).padStart(2)}. ${t.team.padEnd(25)} ${String(t.points).padStart(3)} pts | ${t.wins}W ${t.draws}D ${t.losses}L | GF:${t.goalsFor} GA:${t.goalsAgainst} GD:${gdStr}`;
  });

  return [`${compLabel} ${params.season} Standings (calculated from results)`, "=".repeat(70), ...lines].join("\n");
}

export function compareTeams(params: { team1: string; team2: string; season?: number }): string {
  const { matches } = getDataStore();

  const getTeamMatches = (team: string) =>
    matches.filter((m) => {
      const involved = teamMatches(m.homeTeam, team) || teamMatches(m.awayTeam, team);
      if (!involved) return false;
      if (params.season && m.season !== params.season) return false;
      if (m.competition !== "brasileirao" && m.competition !== "historical") return false;
      return true;
    });

  const m1 = getTeamMatches(params.team1);
  const m2 = getTeamMatches(params.team2);

  if (m1.length === 0 && m2.length === 0) {
    return `No data found for either ${params.team1} or ${params.team2}.`;
  }

  const s1 = computeStats(m1, params.team1);
  const s2 = computeStats(m2, params.team2);

  const seasonStr = params.season ? ` (${params.season})` : "";

  const formatRow = (label: string, v1: string | number, v2: string | number) =>
    `${label.padEnd(20)} ${String(v1).padStart(10)}   ${String(v2).padStart(10)}`;

  return [
    `Team Comparison${seasonStr}`,
    formatRow("", params.team1, params.team2),
    "-".repeat(45),
    formatRow("Matches", s1.matches, s2.matches),
    formatRow("Points", s1.points, s2.points),
    formatRow("Wins", s1.wins, s2.wins),
    formatRow("Draws", s1.draws, s2.draws),
    formatRow("Losses", s1.losses, s2.losses),
    formatRow("Goals For", s1.goalsFor, s2.goalsFor),
    formatRow("Goals Against", s1.goalsAgainst, s2.goalsAgainst),
    formatRow(
      "Goal Diff",
      (s1.goalsFor - s1.goalsAgainst >= 0 ? "+" : "") + (s1.goalsFor - s1.goalsAgainst),
      (s2.goalsFor - s2.goalsAgainst >= 0 ? "+" : "") + (s2.goalsFor - s2.goalsAgainst)
    ),
    formatRow("Win Rate", s1.matches > 0 ? `${((s1.wins / s1.matches) * 100).toFixed(1)}%` : "0%", s2.matches > 0 ? `${((s2.wins / s2.matches) * 100).toFixed(1)}%` : "0%"),
  ].join("\n");
}

export function getBestHomeRecord(params: { season?: number; competition?: string; limit?: number }): string {
  const { matches } = getDataStore();
  const limit = params.limit ?? 10;

  const filtered = matches.filter((m) => {
    if (params.season && m.season !== params.season) return false;
    if (params.competition) {
      const comp = params.competition.toLowerCase();
      if ((comp.includes("brasileirao") || comp.includes("serie a")) && m.competition !== "brasileirao" && m.competition !== "historical") return false;
    } else {
      if (m.competition !== "brasileirao" && m.competition !== "historical") return false;
    }
    return true;
  });

  const homeMap = new Map<string, { wins: number; draws: number; losses: number; matches: number }>();
  for (const m of filtered) {
    const key = m.homeTeam.toLowerCase();
    if (!homeMap.has(key)) homeMap.set(key, { wins: 0, draws: 0, losses: 0, matches: 0, name: m.homeTeam } as any);
    const entry = homeMap.get(key)!;
    entry.matches++;
    if (m.homeGoals > m.awayGoals) entry.wins++;
    else if (m.homeGoals === m.awayGoals) entry.draws++;
    else entry.losses++;
  }

  const teams = Array.from(homeMap.entries())
    .map(([, v]) => v as any)
    .filter((v) => v.matches >= 5)
    .sort((a, b) => {
      const rateB = b.wins / b.matches;
      const rateA = a.wins / a.matches;
      return rateB - rateA;
    });

  const seasonStr = params.season ? ` (${params.season})` : "";
  const lines = teams.slice(0, limit).map((t, i) => {
    const rate = ((t.wins / t.matches) * 100).toFixed(1);
    return `${i + 1}. ${t.name}: ${t.wins}W ${t.draws}D ${t.losses}L (${rate}% win rate, ${t.matches} home matches)`;
  });

  return [`Best Home Records${seasonStr}`, ...lines].join("\n");
}

File: experiments-local/experiment-mu-sonnet-claudecode/runs/agent=claude-code_language=csharp_model=sonnet-4.6_tooling=none/rep2/src/BrazilianSoccerMcp/Tools/TeamTools.cs
using System.ComponentModel;
using System.Globalization;
using System.Text;
using BrazilianSoccerMcp.Services;
using ModelContextProtocol.Server;

namespace BrazilianSoccerMcp.Tools;

[McpServerToolType]
public sealed class TeamTools
{
    private readonly DataService _data;

    public TeamTools(DataService data) => _data = data;

    private static bool CompetitionMatches(string competition, string filter)
    {
        var comparer = CultureInfo.InvariantCulture.CompareInfo;
        return comparer.IndexOf(competition, filter, CompareOptions.IgnoreCase | CompareOptions.IgnoreNonSpace) >= 0;
    }

    [McpServerTool]
    [Description("Get win/loss/draw statistics for a team in a given season and/or competition.")]
    public string GetTeamStats(
        [Description("Team name (e.g. 'Flamengo', 'Palmeiras', 'Corinthians').")] string team,
        [Description("Season year (e.g. 2023). If not specified, returns all-time stats.")] int? season = null,
        [Description("Competition filter (e.g. 'Brasileirao', 'Copa do Brasil'). Leave empty for all.")] string? competition = null,
        [Description("Include home vs away breakdown.")] bool homeAwayBreakdown = false)
    {
        var matches = _data.Matches
            .Where(m =>
                (TeamNameNormalizer.Matches(m.HomeTeam, team) || TeamNameNormalizer.Matches(m.AwayTeam, team))
                && m.HomeGoal.HasValue && m.AwayGoal.HasValue)
            .AsEnumerable();

        if (season.HasValue)
            matches = matches.Where(m => m.Season == season);

        if (!string.IsNullOrWhiteSpace(competition))
            matches = matches.Where(m => CompetitionMatches(m.Competition, competition));

        var matchList = matches.ToList();

        if (matchList.Count == 0)
            return $"No match data found for '{team}'{(season.HasValue ? $" in {season}" : "")}.";

        int played = 0, wins = 0, draws = 0, losses = 0, goalsFor = 0, goalsAgainst = 0;
        int homePlayed = 0, homeWins = 0, homeDraws = 0, homeLosses = 0;
        int awayPlayed = 0, awayWins = 0, awayDraws = 0, awayLosses = 0;

        foreach (var m in matchList)
        {
            bool isHome = TeamNameNormalizer.Matches(m.HomeTeam, team);
            int myGoals = isHome ? m.HomeGoal!.Value : m.AwayGoal!.Value;
            int oppGoals = isHome ? m.AwayGoal!.Value : m.HomeGoal!.Value;

            played++;
            goalsFor += myGoals;
            goalsAgainst += oppGoals;

            if (myGoals > oppGoals) { wins++; if (isHome) homeWins++; else awayWins++; }
            else if (myGoals < oppGoals) { losses++; if (isHome) homeLosses++; else awayLosses++; }
            else { draws++; if (isHome) homeDraws++; else awayDraws++; }

            if (isHome) homePlayed++;
            else awayPlayed++;
        }

        int points = wins * 3 + draws;
        double winRate = played > 0 ? (double)wins / played * 100 : 0;

        var sb = new StringBuilder();
        sb.AppendLine($"{team} Statistics{(season.HasValue ? $" ({season})" : " (all seasons)")}:");
        if (!string.IsNullOrWhiteSpace(competition)) sb.AppendLine($"Competition: {competition}");
        sb.AppendLine();
        sb.AppendLine($"Matches: {played} | Points: {points}");
        sb.AppendLine($"Wins: {wins} | Draws: {draws} | Losses: {losses}");
        sb.AppendLine($"Goals For: {goalsFor} | Goals Against: {goalsAgainst} | Goal Diff: {goalsFor - goalsAgainst}");
        sb.AppendLine($"Win Rate: {winRate:F1}%");

        if (homeAwayBreakdown)
        {
            sb.AppendLine();
            sb.AppendLine($"Home record ({homePlayed} matches): {homeWins}W {homeDraws}D {homeLosses}L");
            sb.AppendLine($"Away record ({awayPlayed} matches): {awayWins}W {awayDraws}D {awayLosses}L");
        }

        return sb.ToString();
    }

    [McpServerTool]
    [Description("Compare two teams head-to-head statistics and overall performance side by side.")]
    public string CompareTeams(
        [Description("First team name.")] string team1,
        [Description("Second team name.")] string team2,
        [Description("Season year (optional).")] int? season = null)
    {
        var h2hMatches = _data.Matches
            .Where(m =>
                ((TeamNameNormalizer.Matches(m.HomeTeam, team1) && TeamNameNormalizer.Matches(m.AwayTeam, team2)) ||
                 (TeamNameNormalizer.Matches(m.HomeTeam, team2) && TeamNameNormalizer.Matches(m.AwayTeam, team1)))
                && m.HomeGoal.HasValue && m.AwayGoal.HasValue)
            .ToList();

        if (season.HasValue)
            h2hMatches = h2hMatches.Where(m => m.Season == season).ToList();

        int t1Wins = 0, t2Wins = 0, draws = 0;
        foreach (var m in h2hMatches)
        {
            bool t1IsHome = TeamNameNormalizer.Matches(m.HomeTeam, team1);
            int t1Goals = t1IsHome ? m.HomeGoal!.Value : m.AwayGoal!.Value;
            int t2Goals = t1IsHome ? m.AwayGoal!.Value : m.HomeGoal!.Value;
            if (t1Goals > t2Goals) t1Wins++;
            else if (t2Goals > t1Goals) t2Wins++;
            else draws++;
        }

        var sb = new StringBuilder();
        sb.AppendLine($"Comparison: {team1} vs {team2}");
        sb.AppendLine();

        if (h2hMatches.Count > 0)
        {
            sb.AppendLine($"Head-to-Head ({h2hMatches.Count} matches):");
            sb.AppendLine($"  {team1}: {t1Wins} wins | {team2}: {t2Wins} wins | Draws: {draws}");
            sb.AppendLine();
        }

        foreach (var (teamName, label) in new[] { (team1, "Team 1"), (team2, "Team 2") })
        {
            var teamMatches = _data.Matches
                .Where(m =>
                    (TeamNameNormalizer.Matches(m.HomeTeam, teamName) || TeamNameNormalizer.Matches(m.AwayTeam, teamName))
                    && m.HomeGoal.HasValue && m.AwayGoal.HasValue
                    && (!season.HasValue || m.Season == season))
                .ToList();

            int wins = 0, d = 0, losses = 0, gf = 0, ga = 0;
            foreach (var m in teamMatches)
            {
                bool isHome = TeamNameNormalizer.Matches(m.HomeTeam, teamName);
                int myG = isHome ? m.HomeGoal!.Value : m.AwayGoal!.Value;
                int oppG = isHome ? m.AwayGoal!.Value : m.HomeGoal!.Value;
                gf += myG; ga += oppG;
                if (myG > oppG) wins++;
                else if (myG < oppG) losses++;
                else d++;
            }

            sb.AppendLine($"{teamName} overall{(season.HasValue ? $" ({season})" : "")}:");
            sb.AppendLine($"  Matches: {teamMatches.Count} | W: {wins} D: {d} L: {losses} | GF: {gf} GA: {ga} GD: {gf - ga}");
        }

        return sb.ToString();
    }

    [McpServerTool]
    [Description("Find all competitions a team has participated in within the dataset.")]
    public string GetTeamCompetitions(
        [Description("Team name to look up.")] string team)
    {
        var teamMatches = _data.Matches
            .Where(m =>
                TeamNameNormalizer.Matches(m.HomeTeam, team) ||
                TeamNameNormalizer.Matches(m.AwayTeam, team))
            .ToList();

        if (teamMatches.Count == 0)
            return $"No matches found for team '{team}'.";

        var competitions = teamMatches
            .GroupBy(m => m.Competition)
            .Select(g => new { Competition = g.Key, Count = g.Count(), Seasons = g.Select(m => m.Season).Where(s => s.HasValue).Distinct().OrderBy(s => s).ToList() })
            .OrderByDescending(g => g.Count)
            .ToList();

        var sb = new StringBuilder();
        sb.AppendLine($"Competitions for {team} ({teamMatches.Count} total matches):");
        sb.AppendLine();

        foreach (var c in competitions)
        {
            var seasonsStr = c.Seasons.Count > 0 ? $" | Seasons: {string.Join(", ", c.Seasons)}" : "";
            sb.AppendLine($"- {c.Competition}: {c.Count} matches{seasonsStr}");
        }

        return sb.ToString();
    }
}
